[debug-info] More fixes for entry values in SelectionDAG.

gottesmm opened the pull request:
This PR contains a few fixes for selection DAG around entry values. I go give a brief description of each patch below:

1. In the first patch, I add a new form of SDDbgOperand that forces an entry value to be emitted rather than relying on heuristics in VarLocBasedImpl. I also wire up InstrEmitter so that if it sees such a SDDbgOperand, it knows what to do.
2. In patch 2, I teach selection dag how to use this new SDDbgOperand to emit swift async context params always as entry values.
3. In patch 3, I change the current code that is used to emit frame var llvm.dbg.addr to pass false for the "parameter" flag on DAG.addDbgValue. The reason why I am doing this is that the name of the flag parameter (called "is parameter") is a misnomer. If set to true, there is an implication that one is creating a dbg value for a byval argument that is intended to have a dbg.declare. As a result the implicit will hoist the relevant debug instruction to the entry block, incorrect behavior if we have a dbg.addr. Instead, we just set that to false and we keep the relevant DBG_VALUE in the appropriate place. This importantly fixed a bug where we were not handling debug info of owned inout arguments correctly.
4. I added some llvm-ir llvm-dwarfdump tests to validate our implementation based off performing -emit-ir on some swift.

gottesmm committed:
[debug-info] Teach SelectionDAG how to always emit swift async context arguments as entry values.

I also added a small fix in VarLocBasedImpl to not try to change DEBUG_VAR that
are already entry value into entry value.

gottesmm committed:
[debug-info] Stop hoisting DBG_VALUE from llvm.dbg.addr on frame index or allocas.

The parameter that I am changing to false here is called "parameter", but it is
actually misnamed. It really implies semantically that we have a byval parameter
that we want to emit dbg.declare upon. This information is then used to justify
hoisting the DBG_VALUE to the front of a function. This is correct for
llvm.dbg.declare but is not correct for llvm.dbg.addr since is always a control
dependent intrinsic so must be left in position, rather than hoisted.

gottesmm committed:
[debug-info] Add llvm-ir tests that exercise codegen debug info around Swift move function codegen.

gottesmm committed:
Reorder lines in a test slightly.

I think this is just a harmless change that returns the codegen to a slightly
older form.

gottesmm commented:
@swift-ci test

gottesmm commented on llvm/test/DebugInfo/X86/dbg-addr-dse.ll:
```diff
@@ -46,8 +46,8 @@ entry:
-; ASM: #DEBUG_VALUE: f:x <- [DW_OP_plus_uconst [[OFF_X:[0-9]+]]] [$rsp+0]
-; ASM: movl    %ecx, [[OFF_X]](%rsp)
+; ASM: movl    %ecx, [[OFF_X:[0-9]+]](%rsp)
+; ASM: #DEBUG_VALUE: f:x <- [DW_OP_plus_uconst [[OFF_X]]] [$rsp+0]
```
@adrian-prantl do you have any concerns about this? I looked in the history and the code used to look like this. Just got re-ordered slightly.

adrian-prantl replied:
IIUC, the new order is more precise.

gottesmm commented:
apple/swift#42125

@swift-ci test

gottesmm commented:
https://github.com/apple/swift/pull/42125

@swift-ci test

adrian-prantl commented on llvm/lib/CodeGen/LiveDebugValues/VarLocBasedImpl.cpp:
```diff
@@ -2196,24 +2196,28 @@ bool VarLocBasedLDV::ExtendRanges(MachineFunction &MF,
+        // If our instruction is not an entry value yet, make it an entry value.
+        if (!MI.getDebugExpression()->isEntryValue()) {
+          MI.getOperand(3).setMetadata(DIExpression::prepend(
+              MI.getDebugExpression(), DIExpression::EntryValue));
```
This looks good.

adrian-prantl commented on lldb/test/API/lang/swift/variables/move_function/TestSwiftMoveFunction.py:
```diff
@@ -283,8 +280,7 @@ def do_check_addressonly_var_arg_test(self):
-        #self.assertGreater(varK.unsigned, 0, "varK not initialized")
-        self.assertIsNone(varK.value, "K is live but was moved?!")
+        self.assertGreater(varK.unsigned, 0, "varK not initialized")
 
```
This looks good.

adrian-prantl replied:
Have you tested this on arm64, too?

gottesmm replied:
I haven't tested this on arm64, specifically. It might be working now, I am not sure. I assumed I would have to add GISEL support for this since we are changing to that in the future.

adrian-prantl commented on llvm/lib/CodeGen/LiveDebugValues/VarLocBasedImpl.cpp:
```diff
@@ -2196,24 +2196,28 @@ bool VarLocBasedLDV::ExtendRanges(MachineFunction &MF,
   if (AsyncDbgValues.size()) {
     // Make sure the async entry values are at the very start.
-    auto InsertPt = First_MBB.getFirstNonDebugInstr();
+    auto InsertPt = FirstMBB.getFirstNonDebugInstr();
```
This looks good if it matches upstream.

gottesmm replied:
I'm going to undo this. I did the change since my editor was complaining about this and I prepared the patch originally for llvm.org where a patch changing First_MBB -> FirstMBB wouldn't cause issues.

adrian-prantl commented on llvm/lib/CodeGen/LiveDebugValues/VarLocBasedImpl.cpp:
```diff
@@ -2196,24 +2196,28 @@ bool VarLocBasedLDV::ExtendRanges(MachineFunction &MF,
-        MI.getOperand(3).setMetadata(DIExpression::prepend(
-            MI.getDebugExpression(), DIExpression::EntryValue));
+        // If our instruction is not an entry value yet, make it an entry value.
+        if (!MI.getDebugExpression()->isEntryValue()) {
```
You'll also need to make the same change in `InstrRefBasedImpl.cpp`.

gottesmm replied:
How do I test that?

gottesmm replied:
I added the code in the place that seemed the most similar.

adrian-prantl commented on llvm/lib/CodeGen/SelectionDAG/InstrEmitter.cpp:
```diff
@@ -721,9 +721,15 @@ void InstrEmitter::AddDbgValueLocationOps(
+    }
+    case SDDbgOperand::ENTRYVALUEREG: {
+      // We will add our entry value to the expression in our caller.
+      MIB.addReg(Op.getEntryValueReg());
```
This LGTM, but we need to start a discussion about upstreaming this.

gottesmm replied:
I am fine just upstreaming this. The problem is I need tests to upstream and my tests seem to not be working in OSS as I mentioned due to missing cherry-picks.

adrian-prantl commented on llvm/lib/CodeGen/SelectionDAG/SDNodeDbgValue.h:
```diff
@@ -107,6 +121,7 @@ class SDDbgOperand {
     const Value *Const; ///< Valid for constants.
     unsigned FrameIx;   ///< Valid for stack objects.
     unsigned VReg;      ///< Valid for registers.
+    unsigned EntryValueReg; ///< Valid for physiucal regs that need entry value.
```
physical

adrian-prantl commented on llvm/lib/CodeGen/SelectionDAG/SDNodeDbgValue.h:
```diff
@@ -30,10 +30,13 @@ class raw_ostream;
+    CONST = 1,         ///< Value is a constant.
+    FRAMEIX = 2,       ///< Value is contents of a stack location.
+    VREG = 3,          ///< Value is a virtual register.
+    ENTRYVALUEREG = 4, ///< Value is a physical register that should be
```
Do we really need a new Kind or would it be sufficient to append the EntryValue to the `DIExpression` when we create the `SDDbgNode` and look for VREG+EntryValue in the code?

gottesmm replied:
I tried this and it worked. So I am getting rid of this code. Great idea!

adrian-prantl commented on llvm/test/DebugInfo/X86/swift-move-function-debuginfo-async.ll:
```diff
@@ -0,0 +1,1895 @@
+;; RUN: llc -O0 -mtriple=x86_64-apple-darwin %s -o %t -filetype=obj
+;; RUN: llvm-dwarfdump --show-children %t | FileCheck --check-prefix=DWARF %s
+
+source_filename = "move_function_dbginfo_async.ll"
```
Is ther anything we could do here (llvm-extract?) to make this test shorter? This looks very hard to maintain.

gottesmm replied:
I'll try to reduce it. The reason why I wanted to include most of this is that various of the subroutines test specific behavior when they are processed. I am going to try to llvm-extract just those.

adrian-prantl commented on llvm/test/DebugInfo/X86/swift-move-function-debuginfo.ll:
```diff
@@ -0,0 +1,2329 @@
+!529 = !DILocation(line: 721, column: 11, scope: !511)
+!530 = !DILocation(line: 722, column: 5, scope: !512)
+!531 = !DILocation(line: 718, column: 5, scope: !512)
+!532 = !DILocation(line: 723, column: 5, scope: !513)
```
Same here.

gottesmm commented:
https://github.com/apple/swift/pull/42125

@swift-ci test

gottesmm commented:
lldb tests passed. Going to push the branch with the requested changes.

gottesmm commented:
https://github.com/apple/swift/pull/42125

@swift-ci test

gottesmm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -5716,6 +5720,49 @@ bool SelectionDAGBuilder::EmitFuncArgumentDbgValue(
+    // Then add the entry value to our expression.
+    Expr = DIExpression::prepend(Expr, DIExpression::EntryValue);
+
+    // And create the vreg dbg value with an entry value expression prepended.
```
@adrian-prantl I took your advice and just added the entry value this way.

gottesmm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -5716,6 +5720,49 @@ bool SelectionDAGBuilder::EmitFuncArgumentDbgValue(
     return false;
 
+  // If we had a dbg addr, just emit it here.
+  if (Kind == FuncArgumentDbgValueKind::Addr) {
```
here

gottesmm committed:
[debug-info] As an NFC commit, refactor EmitFuncArgumentDbgValue so that it can be extended to support llvm.dbg.addr.

The reason why I am making this change is that before this commit,
EmitFuncArgumentDbgValue relied on a boolean flag IsDbgDeclare both to signal
that a DBG_VALUE should be made to be indirect /and/ that the original intrinsic
was a dbg.declare. This is no longer always true if we add support for handling
dbg.addr since we will have an indirect DBG_VALUE that is a different intrinsic
from dbg.declare.

With that in mind, in this NFC patch, we prepare for future fixes by introducing
a 3 case-enum argument to EmitFuncArgumentDbgValue that allows the caller to
explicitly specify how the argument's DBG_VALUE should be emitted. This then
allows us to turn the indirect checks into a != FuncArgumentDbgValueKind::Value
and prepare us for a future where we add support here for llvm.dbg.addr
directly.

rdar://83957028

Reviewed By: aprantl

Differential Revision: https://reviews.llvm.org/D122945

(cherry picked from commit e24f5348798605a799c63ff09169d177d262cd37)

gottesmm commented:
AdrianP asked me to upstream part of this as a NFC commit on llvm.org. That is here: https://reviews.llvm.org/D122945

gottesmm commented:
That being said... it is good to see everything passed!

gottesmm commented:
https://github.com/apple/swift/pull/42125

@swift-ci test
